Source/WebKit/UIProcess/API/glib/WebKitImage.cpp

```diff
@@ -0,0 +1,477 @@
+ * Returns: the image height
+ *
+ * Since: 2.52
+ */
+int webkit_image_get_height(WebKitImage* image)
+{
+    g_return_val_if_fail(WEBKIT_IS_IMAGE(image), 0);
+    return image->priv->height;
+}
+
+/**
+ * webkit_image_get_stride:
+ * @image: a #WebKitImage
+ *
+ * Get the @image stride
+ *
+ * Returns: the image stride
+ *
+ * Since: 2.52
+ */
+guint webkit_image_get_stride(WebKitImage* image)
+{
+    g_return_val_if_fail(WEBKIT_IS_IMAGE(image), 0);
+    return image->priv->stride;
+}
+
+/**
+ * webkit_image_as_rgba_bytes:
+ * @image: a #WebKitImage
+ *
+ * Get the @image bytes in their native format
+ *
+ * Returns: (transfer none): a #GBytes
+ *
+ * Since: 2.52
+ */
+GBytes* webkit_image_as_rgba_bytes(WebKitImage* image)
+{
+    g_return_val_if_fail(WEBKIT_IS_IMAGE(image), nullptr);
+
+    if (!image->priv->bytes) {
+        ASSERT(image->priv->createBytesFn);
+        image->priv->bytes = image->priv->createBytesFn(image);
+    }
+    ASSERT(image->priv->bytes);
+
+    return image->priv->bytes.get();
+}
+
+#if PLATFORM(GTK)
+/**
+ * webkit_image_to_gdk_texture:
+ * @image: a #WebKitImage
+ *
+ * Get the @image as a #GdkTexture
+ *
+ * Returns: (transfer full): a #GdkTexture
+ *
+ * Since: 2.52
+ */
+GdkTexture* webkit_image_to_gdk_texture(WebKitImage *image)
+{
+    g_return_val_if_fail(WEBKIT_IS_IMAGE(image), nullptr);
+
+    auto* bytes = webkit_image_as_rgba_bytes(image);
+
+    return gdk_memory_texture_new(image->priv->width, image->priv->height, GDK_MEMORY_DEFAULT, bytes, image->priv->stride);
+}
+#endif
+
+static guint webkit_image_hash(GIcon*);
+static gboolean webkit_image_equal(GIcon*, GIcon*);
+
+static GInputStream* webkit_image_load(GLoadableIcon*, int, char**, GCancellable*, GError**);
+static void webkit_image_load_async(GLoadableIcon*, int, GCancellable*, GAsyncReadyCallback, gpointer);
+static GInputStream* webkit_image_load_finish(GLoadableIcon*, GAsyncResult*, char**, GError**);
+
+static void webkit_image_gicon_interface_init(GIconIface* iface)
+{
+    iface->hash = webkit_image_hash;
+    iface->equal = webkit_image_equal;
+}
+
+static void webkit_image_gloadable_icon_interface_init(GLoadableIconIface* iface)
+{
+    iface->load = webkit_image_load;
+    iface->load_async = webkit_image_load_async;
+    iface->load_finish = webkit_image_load_finish;
+}
+
+guint webkit_image_hash(GIcon* icon)
+{
+    g_return_val_if_fail(WEBKIT_IS_IMAGE(icon), 0);
+
+    auto* image = WEBKIT_IMAGE(icon);
+
+    Hasher hasher;
+
+    auto* bytes = webkit_image_as_rgba_bytes(image);
+    gsize size = 0;
+    const guint8* data = static_cast<const guint8*>(g_bytes_get_data(bytes, &size));
+    gsize rowBytes = image->priv->width * static_cast<uint8_t>(ImageFormatBPP::RGBA8);
+
+    ASSERT(size >= image->priv->stride * (image->priv->height - 1) + rowBytes);
+
+    for (int y = 0; y < image->priv->height; ++y) {
+        auto span = std::span<const uint8_t>(data + y * image->priv->stride, rowBytes);
+        WTF::add(hasher, span);
+    }
```

Comment:
I think this could use `consumeSpan()` and `span::first()` to avoid the pointer arithmetic altogether:

```suggestion
    auto rowData = span(bytes);
    for (int row = 0; row < image->priv->height; ++row) {
        WTF::add(hasher, rowData.first(rowBytes));
        rowData = consumeSpan(rowData, image->priv->stride);
    }
```